### ISSUE TITLE
Port already in use when attempting to inject sidecar

### ISSUE THREAD

**Issue by petenorth** [istio/istio#10765]
**Describe the bug**
Unable to inject sidecar pod for micronaut based app

**Expected behavior**
A pod with two containers remains at ready state

**Steps to reproduce the bug**
    kubectl apply -f https://raw.githubusercontent.com/petenorth/micronaut-tutorial-1/master/kubernetes/deployments.yaml
    kubectl get deployment books --export=true -o yaml > books.yaml
    istioctl kube-inject -f books.yaml | kubectl apply -f -
    kubectl get pod -w

If I replace the image with a spring boot app  (petenorth/gs-reactive-rest-service updating the health check to /actuator/health and the port to 8080) then injection works. 

The error in the logs  is

    16:20:41.535 [main] INFO  i.m.context.env.DefaultEnvironment - Established active environments [cloud, k8s]
    16:20:43.043 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: 
    Unable to start Micronaut server on port: 8082 io.micronaut.http.server.exceptions.ServerStartupException: Unable to start Micronaut server on port: 8082
    at io.micronaut.http.server.netty.NettyHttpServer.bindServerToHost(NettyHttpServer.java:409)
    at io.micronaut.http.server.netty.NettyHttpServer.start(NettyHttpServer.java:303)
    at io.micronaut.http.server.netty.NettyHttpServer.start(NettyHttpServer.java:107) 
    at io.micronaut.runtime.Micronaut.lambda$start$1(Micronaut.java:73)
    at java.util.Optional.ifPresent(Optional.java:159)
    at io.micronaut.runtime.Micronaut.start(Micronaut.java:71)
    at io.micronaut.runtime.Micronaut.run(Micronaut.java:271)
    at io.micronaut.runtime.Micronaut.run(Micronaut.java:257)
    at example.micronaut.Application.main(Application.groovy:9)
    Caused by: java.net.BindException: Address already in use
	... 9 common frames omitted

**Version**
istio: 1.0.5
minikube version: v0.32.0
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.3", GitCommit:"2bba0127d85d5a46ab4b778548be28623b32d0b0", GitTreeState:"clean", BuildDate:"2018-05-21T09:17:39Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.4", GitCommit:"f49fa022dbe63faafd0da106ef7e05a29721d3f1", GitTreeState:"clean", BuildDate:"2018-12-14T06:59:37Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}

**Installation**

    minikube start --vm-driver=xhyve --memory 5140
    #IMPORTANT version 2.12.0 of helm doesn't work
    alias helm="docker run -ti --rm -v $(pwd):/apps:rw -v ~/.kube:/root/.kube -v ~/.minikube:$(echo ~)/.minikube alpine/helm:2.11.0"
    #is the service account needed etc. when the default minikube install doesn't have RBAC enabled?
    kubectl apply -f install/kubernetes/helm/helm-service-account.yaml
    helm init --service-account tiller
    helm install install/kubernetes/helm/istio --name istio-minimal --namespace istio-system \
      --set security.enabled=false \
      --set ingress.enabled=false \
      --set gateways.istio-ingressgateway.enabled=false \
      --set gateways.istio-egressgateway.enabled=false \
      --set galley.enabled=false \
      --set sidecarInjectorWebhook.enabled=false \
      --set mixer.enabled=false \
      --set prometheus.enabled=false \
      --set global.proxy.envoyStatsd.enabled=false \
      --set pilot.sidecar=false
    kubectl get pods -n istio-system -w

**Environment**
minikube on macos

**Cluster state**
N/A

**Comment by vadimeisenbergibm** [comment]
Can you check who listens to that port in your container?

E.g.: `netstat -plnt`.

I tried:
1. Inside the application container:
```
root@sleep-76df4f989c-ncps2:/# netstat -plnt         
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:15090           0.0.0.0:*               LISTEN      -               
tcp        0      0 127.0.0.1:15000         0.0.0.0:*               LISTEN      -               
tcp        0      0 0.0.0.0:15001           0.0.0.0:*               LISTEN      -        
```

1. Inside the `istio-proxy` container:
```
istio-proxy@sleep-76df4f989c-ncps2:/$ netstat -plnt
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:15090           0.0.0.0:*               LISTEN      18/envoy        
tcp        0      0 127.0.0.1:15000         0.0.0.0:*               LISTEN      18/envoy        
tcp        0      0 0.0.0.0:15001           0.0.0.0:*               LISTEN      18/envoy     
```

Note that PID/Program name of Envoy inside `istio-proxy` container.

**Comment by petenorth** [comment]
I have created a test image 

https://hub.docker.com/r/petenorth/test

The image has everything (including the java source for SocketQuery.java). The Dockerfile for this image is

    FROM openjdk:8u171-alpine3.7
    RUN apk --no-cache add curl
    COPY build/libs/*-all.jar books.jar
    COPY run.sh run.sh
    COPY SocketQuery.class SocketQuery.class
    COPY SocketQuery.java SocketQuery.java
    RUN chmod +x run.sh
    RUN pwd
    RUN ls -l
    CMD ./run.sh

When I look at the container logs I get (the two lines immediately after the netstat output are from the SocketQuery java class which mimics the code at NettyHttpServer.java:409)

    Active Internet connections (only servers)
    Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
    tcp        0      0 0.0.0.0:15090           0.0.0.0:*               LISTEN      -
    tcp        0      0 127.0.0.1:15000         0.0.0.0:*               LISTEN      -
    tcp        0      0 0.0.0.0:15001           0.0.0.0:*               LISTEN      -
    host used is books-5c8775bbf6-q5jnn/172.17.0.9
    isTcpPortAvailable = false
    16:24:02.901 [main] INFO  i.m.context.env.DefaultEnvironment - Established active environments:[cloud, k8s]
    16:24:04.332 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: Unable to start Micronaut server on port: 8082
    io.micronaut.http.server.exceptions.ServerStartupException: Unable to start Micronaut server on port: 8082
	at io.micronaut.http.server.netty.NettyHttpServer.bindServerToHost(NettyHttpServer.java:409)
	at io.micronaut.http.server.netty.NettyHttpServer.start(NettyHttpServer.java:303)
	at io.micronaut.http.server.netty.NettyHttpServer.start(NettyHttpServer.java:107)
	at io.micronaut.runtime.Micronaut.lambda$start$1(Micronaut.java:73)
	at java.util.Optional.ifPresent(Optional.java:159)
	at io.micronaut.runtime.Micronaut.start(Micronaut.java:71)
	at io.micronaut.runtime.Micronaut.run(Micronaut.java:271)
	at io.micronaut.runtime.Micronaut.run(Micronaut.java:257)
	at example.micronaut.Application.main(Application.groovy:9)
    Caused by: java.net.BindException: Address already in use
	... 9 common frames omitted

**Comment by vadimeisenbergibm** [comment]
I do not see that the port 8082 is in use by Envoy or some other app. Does your app runs without Istio? Does it run with Istio on some cluster other than minikube?

**Comment by petenorth** [comment]
Yes the application runs without istio:

    kubectl apply -f https://raw.githubusercontent.com/petenorth/micronaut-tutorial-1/master/kubernetes/deployments.yaml

results in three deployments with pods running successfully. I have deployed the test image without injection and see the following

    Active Internet connections (servers and established)
    Proto Recv-Q Send-Q Local Address           Foreign Address         State       
    Active UNIX domain sockets (servers and established)
    Proto RefCnt Flags       Type       State         I-Node Path
    Active Internet connections (only servers)
    Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
    host used is books-78d6df9589-lh74q/172.17.0.9
    isTcpPortAvailable = true
    11:06:17.001 [main] INFO  i.m.context.env.DefaultEnvironment - Established active environments: [cloud, k8s]
    11:06:18.720 [main] INFO  io.micronaut.runtime.Micronaut - Startup completed in 1711ms. Server Running: http://books-78d6df9589-lh74q:8082  

i.e. no socket connections prior to application startup.

**Comment by vadimeisenbergibm** [comment]
Ok, so the second question - Does it run with Istio on some cluster other than minikube?

**Comment by petenorth** [comment]
A framework code change could fix this but it isn't clear why this change is required, anyway see

https://github.com/micronaut-projects/micronaut-core/issues/1081#issuecomment-451980818

The framework check for whether the TCP port is available seems reasonable.

**Comment by vadimeisenbergibm** [comment]
My guess is it is somehow related to the initialization of the container, somehow a port becomes unavailable for some period of time. Maybe the framework should wait for the port to become available for some time, before giving up.

**Comment by petenorth** [comment]
This is fixed in Micronaut 1.0.3 (yet to be released) - tracked by the issue mentioned in this comment https://github.com/istio/istio/issues/10765#issuecomment-451982815
